### PR TITLE
Migrated integration_flutter_test/embedder (scenic integration test) from fuchsia.git

### PR DESCRIPTION
**Pull request by richkadel**
The tests use upgraded flutter/repo build rules, implemented in previously merged PRs, starting with #27996, to compile C++ and Dart tests as Fuchsia components, with dependencies on both C++ and Dart FIDL bindings, and on additional Dart APIs included in the Fuchsia SDK.

The tests are rewritten, from using Flutter APIs in fuchsia.git, to using only `dart_ui` (since `flutter/engine` tests cannot depend on the downstream `flutter/flutter` APIs), but the basic test program shells have been compiled and successfully run after publishing the test component to a running Fuchsia instance. (See `//flutter/shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/README.md`.)

Now that the rewritten tests are complete and operational, future integration issues between Flutter and Fuchsia (e.g., Scenic) can be caught directly from `flutter/engine` builds, instead of being caught only from Fuchsia builds, which typically uses a version of Flutter that is at least 2 days old.

Note this PR also migrates `dart_library` dependencies from `flutter/buildroot` (secondary/third_party/... packages) to a central flutter fuchsia `BUILD.gn` file, so the `secondary` `BUILD.gn` files that originally defined those `dart_library` targets can now be removed.

*List which issues are fixed by this PR. You must list at least one issue.*

Fixes: flutter/flutter#90286

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [ ] All existing and new tests are passing.
- [ ] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Comment by abarth** [comment]
Please do not land this pull request. It contains files from the Fuchsia repository that are not support for use outside the Fuchsia repository.

**Comment by richkadel** [comment]
Agreed. Thanks Adam. This _Draft_ is not intended to be landed as-is.

**Comment by richkadel** [comment]
Proceeding along the agreed plan.

Not quite there yet, but I was able to replace/remove most of the dependencies so far.

TestWithEnvironment (and required dependencies, minus a pluggable gtest) is still in this PR for now, but will also be removed and migrated to SDK, once formally approved (and before landing this PR, or a derivative).

NOTE: Dart APIs were in the Core SDK but not in GN SDK. The benefits of switching to GN SDK were not enough to justify switching, so I reverted to using the Core SDK, and only had to duplicate 3 .gni files from the GN SDK. (I believe this is in line with recommendations so far.)

I do still need a better way to generate `dart_library` target outputs from a non-BUILD.gn list. For now, to build, I manually created BUILD.gn files for required libraries that needed them (Dart APIs from common), but I should be able to generate these from the JSON meta/manifest data.

File: shell/platform/fuchsia/dart/BUILD.gn
Comment (akbiggs) [inline shell/platform/fuchsia/dart/BUILD.gn:150]: What is using these new Dart dependencies? I don't see references to them in the test .dart code that was added. Why does adding tests increase the dependencies of our platform in general?

**Review by akbiggs — COMMENTED** [review]
Still reviewing, just a pass through the first few files. Sorry for the slow review.

File: shell/platform/fuchsia/dart/BUILD.gn
Comment (akbiggs) [reply shell/platform/fuchsia/dart/BUILD.gn:150]: I see, I missed this part of the description:

> Note this PR also migrates dart_library dependencies from flutter/buildroot (secondary/third_party/... packages) to a central flutter fuchsia BUILD.gn file, so the secondary BUILD.gn files that originally defined those dart_library targets can now be removed.

Might be a good candidate for splitting into a separate pull request, but up to you.

File: shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/README.md
Comment (richkadel) [reply shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/README.md:40]: I don't consider it a bug, exactly. This is pretty standard practice in this situation. You can only have one Scenic on a device with one display. (fuchsia.git has an embeddable scenic component for hermetic tests but it's not currently available OOT.)

Using `fx set core...` (for one) avoids the issue though, and the top of the README uses `core` (not `workstation`) in the example.

File: shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/child-view/child_view.dart
Comment (richkadel) [reply shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/child-view/child_view.dart:8]: You'll need to ignore the test code for now. This is the part that's in progress.

FYI, material.dart is not available to us in flutter/engine. So I have to rewrite the test in dart:ui

File: shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/flutter-embedder-test.cc
Comment (richkadel) [reply shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/flutter-embedder-test.cc:73]: Yeah, that's a good catch. I originally intended to submit this with all of the test code complete, but I'm now doing this in two steps (two PRs), and this really applies to the second one. But the comment is not accurate.

I'll come back to this unresolved comment.

File: shell/platform/fuchsia/dart/BUILD.gn
Comment (richkadel) [reply shell/platform/fuchsia/dart/BUILD.gn:150]: Actually, there are Fuchsia SDK APIs that depend on these. All of the dependencies are in fact in use (when including the integration test).

Comment (richkadel) [reply shell/platform/fuchsia/dart/BUILD.gn:150]: See https://github.com/flutter/engine/pull/26880/files/7b5ce3b1b3504c9ad1fe5330df1ed48784c64485#diff-6378baabb0a230513a31db033afa400fcb2afea278438394af653d525d1b105aR66

File: shell/platform/fuchsia/flutter/integration_flutter_tests/OWNERS
Comment (richkadel) [reply shell/platform/fuchsia/flutter/integration_flutter_tests/OWNERS]: done

File: shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/README.md
Comment (richkadel) [reply shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/README.md]: done

File: shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/child-view/child_view.dart
Comment (richkadel) [reply shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/child-view/child_view.dart:10]: done

File: shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/flutter-embedder-test.cc
Comment (richkadel) [reply shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/flutter-embedder-test.cc]: Actually, as of today, all tests pass. I've removed the `fake-` prefix and tests still pass.

File: shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/README.md
Comment (richkadel) [reply shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/README.md]: done

File: shell/platform/fuchsia/dart/BUILD.gn
Comment (richkadel) [reply shell/platform/fuchsia/dart/BUILD.gn:150]: I could break this PR into 2 or three separate PRs, but if there is no objection, the changes do fit together well enough to keep them together.

File: tools/fuchsia/devshell/serve.sh
Comment (akbiggs) [inline tools/fuchsia/devshell/serve.sh:250]: I messed up this command, it should be "pkgctl repo add". I was running serve.sh incorrectly to test so I didn't notice.

File: tools/fuchsia/devshell/lib/vars.sh
Comment (akbiggs) [inline tools/fuchsia/devshell/lib/vars.sh:41]: yellow WARNING: prefix

File: tools/fuchsia/devshell/serve.sh
Comment (akbiggs) [inline tools/fuchsia/devshell/serve.sh:77]: "sanity check" is non-inclusive, maybe just "check" in this case. The comment also feels like it restates the code so we could probably get rid of it.

File: tools/fuchsia/flutter/flutter_build_config.gni
Comment (akbiggs) [inline tools/fuchsia/flutter/flutter_build_config.gni:28]: Do we need all these extra commented-out configs or should we delete them until we find a need for them in Engine? If we're only planning on using the SDK for making Fuchsia tests run, will there ever be a need for these configs?

**Review by akbiggs — COMMENTED** [review]
This is awesome. The changes to serve.sh help a lot too.

LGTM but should get a review from David/Chase before submission since they have more context on what code is new vs. what code is being ported over from existing folders.

**Comment by richkadel** [comment]
For reviewers, note the comment from @abarth is no longer relevant. The comment raised concerns regarding an early draft that was never intended to be published as is, but served as a starting point only, to expose the early issues trying to port a test like this one. As a result, additional features were added to the Fuchsia SDK, and other known (less critical) issues have been worked around, while we investigate the best path for additional improvements to Fuchsia and the SDK for third-party development.

File: tools/fuchsia/devshell/lib/vars.sh
Comment (richkadel) [reply tools/fuchsia/devshell/lib/vars.sh:41]: good catch!

File: tools/fuchsia/flutter/flutter_build_config.gni
Comment (richkadel) [reply tools/fuchsia/flutter/flutter_build_config.gni:28]: Yes, we will need them, but I didn't get around to validating them yet. I'm not yet as familiar with all of these modes, but this should be updated sooner rather than later. I don't THINK it is critical to publish this PR (but we'll see).

File: tools/fuchsia/devshell/serve.sh
Comment (richkadel) [reply tools/fuchsia/devshell/serve.sh]: done

File: tools/fuchsia/devshell/lib/vars.sh
Comment (richkadel) [reply tools/fuchsia/devshell/lib/vars.sh]: done

File: tools/fuchsia/devshell/serve.sh
Comment (richkadel) [reply tools/fuchsia/devshell/serve.sh]: I rewrote it.

**Comment by richkadel** [comment]
Good to see that CI is fully passing on https://github.com/flutter/engine/pull/26880/commits/35d5a30998a772ce9e82c34c8e522a535eae1a0d

However, the test suites are not yet running the new integration tests.

I've added a new follow-on commit to try to get the new tests added to the test suites (adding an entry in `test_suites.yaml`). The new tests do have some added complexity, including multiple `.far` files, and the requirement to serve the parent-view and child-view packages via the `serve.sh` script, so I'm skeptical this will work out of the box.

**Comment by richkadel** [comment]
except, ugh, more licenses changes to merge. I'll wait until the infra CL merges.

File: shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/BUILD.gn
Comment (arbreng) [inline shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/BUILD.gn:38]: Shouldn't need these; these are internal runner dependencies and the test is using a prebuilt runner

**Review by arbreng — APPROVED** [review]

File: shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/child-view/child_view.dart
Comment (chaselatta) [inline shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/child-view/child_view.dart:9]: small nit: this doesn't need to be async.

**Review by chaselatta — APPROVED** [review]

File: shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/child-view/child_view.dart
Comment (richkadel) [reply shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/child-view/child_view.dart]: done

File: shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/BUILD.gn
Comment (richkadel) [reply shell/platform/fuchsia/flutter/integration_flutter_tests/embedder/BUILD.gn]: done

**Pull request merged**
